### PR TITLE
#9322: Split pad to .hpp and .cpp

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
### Ticket
None

### Problem description
Pad header is not clean, includes device operation header, leaking too much details.

### What's changed
Split .hpp and .cpp
* I tried to minimize .cpp implementation but got stuck and decided to rollback
* I minimized code duplication with macro / this is controversial 
* Ideally macro should go away 

### Checklist
- [ ] Post commit CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/10124968388)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp [hunk 1]
```diff
@@ -5,361 +5,41 @@
 #pragma once
 
 #include "ttnn/tensor/types.hpp"
-#include "ttnn/operations/core/core.hpp"
-
-#include "ttnn/run_operation.hpp"
-
-#include "device/pad_op.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp:12]: this is very good

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp [hunk 2]
```diff
@@ -5,361 +5,41 @@
...
 
 #include <ranges>
 
+#include "ttnn/decorators.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp:11]: but then this header pulls in some heavy things with it

**Review by eyonland — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    const ttnn::Tensor& input_tensor,
+    std::vector<std::pair<uint32_t, uint32_t>> padding,
+    const float value,
+    const bool use_multicore,
+    const std::optional<MemoryConfig>& memory_config_arg) {
+    const int original_rank = input_tensor.get_shape().rank();
+
+    ttnn::Tensor output_tensor;
+    if (input_tensor.storage_type() != StorageType::DEVICE) {
+        if(original_rank == 1) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:128]: I tried to simplify this, but did not succeed. Left as is.

Comment (eyonland) [reply ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:128]: switch (original_rank) {
            case 1: output_tensor = pad_impl<tt::tt_metal::Array1D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 2: output_tensor = pad_impl<tt::tt_metal::Array2D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 3: output_tensor = pad_impl<tt::tt_metal::Array3D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 4: output_tensor = pad_impl<tt::tt_metal::Array4D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 5: output_tensor = pad_impl<tt::tt_metal::Array5D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 6: output_tensor = pad_impl<tt::tt_metal::Array6D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 7: output_tensor = pad_impl<tt::tt_metal::Array7D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 8: output_tensor = pad_impl<tt::tt_metal::Array8D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            default: TT_FATAL("Unsupported tensor rank");
        }

Comment (eyonland) [reply ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:128]:
```
switch (original_rank) {
            case 1: output_tensor = pad_impl<tt::tt_metal::Array1D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 2: output_tensor = pad_impl<tt::tt_metal::Array2D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 3: output_tensor = pad_impl<tt::tt_metal::Array3D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 4: output_tensor = pad_impl<tt::tt_metal::Array4D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 5: output_tensor = pad_impl<tt::tt_metal::Array5D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 6: output_tensor = pad_impl<tt::tt_metal::Array6D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 7: output_tensor = pad_impl<tt::tt_metal::Array7D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 8: output_tensor = pad_impl<tt::tt_metal::Array8D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            default: TT_FATAL("Unsupported tensor rank");
        }
```

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp (hunk 2 above)
Comment (eyonland) [reply ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp:11]: Can this not go down into the cpp?   Not seeing why not.

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp:11]: @eyonland , I wish!

but we define in a header
```
constexpr auto permute = ttnn::register_operation<"ttnn::permute", ttnn::operations::data_movement::ExecutePermute>();
```

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp (hunk 3 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:128]: oh, yes, that should work

**Review by dmakoviichuk-tt — APPROVED** [review]
looks good!

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp [hunk 4]
```diff
@@ -5,361 +5,41 @@
...
-        }
-
-        // Padding always turns the intended shape to the shape with tile padding. For simplicity of the operation
-        output_tensor = ttnn::reshape(output_tensor, ttnn::Shape(padded_shape));
-
-        return output_tensor;
-    }
+    static ttnn::Tensor operator()(uint8_t queue_id,
+                                   const ttnn::Tensor& input_tensor,
+                                   std::vector<std::pair<uint32_t, uint32_t>> padding,
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp:36]: could you pass vector by reference or try std::span to check if it compiles?

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp]: done. But won't touch impl, it modifies padding

**Review by ntarafdar — APPROVED** [review]

**Pull request merged**